### PR TITLE
[codex] Land A3B MTP and DFlash perfmaxx work

### PR DESCRIPTION
**Pull request by Kaden-Schutt**
## Summary

This PR lands the speculative-decode perf work developed across the local branch stack:

- `feat/moe-mtp-a3b`: MoE MTP trunk-spine path, small-N grouped-MoE tile bound fix, device token chain, GPU accept reduction, replay-skip accounting, and Q8 verify-lmhead WMMA routing.
- `feat/moe-dflash-perfmaxx`: DFlash verify-side cleanup, including skipping redundant last-token logits and routing Q8 lmhead verify through gfx12 WMMA.
- `feat/dense-dflash-perfmaxx`: dense/A3B DFlash follow-up routing for HFQ4 lmhead WMMA on gfx12, helper commentary, bench publication notes, and branch-level AGENTS guidance.

## Why

The main goal was to make A3B speculative decode competitive on gfx12 by removing host synchronizations and redundant verify work before considering new kernels. The final validated path is A3B MTP trunk-spine with the calibrated MoE MTP head; DFlash work remains useful as shared infrastructure and as a perf audit trail, but MoE DFlash appears structurally less promising than MoE MTP on the current engine.

## Notable Results

- Published A3B MTP LocalMaxxing run: https://www.localmaxxing.com/runs/cmppmb47l0051pl01lyaup9mc
- Median decode: `259.53 tok/s` on Radeon AI Pro R9700 / gfx12.
- Fixture: Qwen3.6-35B-A3B MQ4-AWQ trunk + MoE MTP head, HumanEval/26 raw prompt.
- Validation run metadata captured in `.codeinsight+research` locally; LocalMaxxing notes include target/head/binary/prompt md5s.

## Implementation Notes

- Fixes the grouped MoE tile bound by aligning runtime and scratch bounds to `MOE_GROUPED_BLOCK_M`, preventing stale/uninitialized expert tile IDs on small verification batches.
- Keeps MQ6/HFQ6 MoE grouped WMMA admission gfx12-scoped; gfx1151 should not inherit that path without a real port and validation.
- Adds a device-resident MTP token chain and GPU-side greedy accept/bonus reduction, reducing per-cycle host sync pressure.
- Adds/gates gfx12 WMMA lmhead verify paths for Q8 and HFQ4-shaped DFlash/MTP verify work.
- Adds `needs_last_token_logits` plumbing so MTP/DFlash verify callers can skip dead scalar logits when they only consume hidden states.

## Validation

Ran before commit:

- `git diff --check`
- `cargo test --features deltanet -p hipfire-arch-qwen35 --lib` — 43 tests passed
- `cargo build --release --features deltanet -p hipfire-runtime --example mtp_only_demo --example dflash_spec_demo`
- pre-commit general coherence gate — no hard errors, report `/tmp/coherence-20260528-150049.md`
- pre-commit DFlash coherence gate — no hard errors, report `/tmp/coherence-dflash-20260528-150118.md`
- pre-commit agentic gate — PASS, report `/tmp/agentic-gate-20260528-150130.md`

## Review Focus

- Check that gfx12-specific dispatch does not accidentally become reachable on gfx11/gfx1151 without a matching kernel.
- Check that the `needs_last_token_logits` default remains safe for non-spec-decode callers.
- Check that the grouped MoE tile bound and sentinel handling are correct for small `K+1` verify batches.
- Check that DFlash/MTP env defaults are appropriately conservative outside validated gfx12 paths.

### CODE REVIEW

**Comment by Kaden-Schutt** [comment]
## Blessed Bench Configs

Pinning the configs we actually blessed during this branch so review/port agents do not accidentally compare against stale trunks, wrong prompts, or adaptive-B variants.

### A3B MTP crown fixture

Published LocalMaxxing run: https://www.localmaxxing.com/runs/cmppmb47l0051pl01lyaup9mc

```bash
./target/release/examples/mtp_only_demo \
  --target /home/kaden/.hipfire/models/qwen3.6-35b-a3b.mq4-awq-mi300x \
  --mtp-head /home/kaden/.hipfire/models/qwen3.6-35b-a3b.moe-mtp-mq4-cvs16384.mtp \
  --prompt-file .codeinsight+research/dense-dflash-perfmaxx/a3b-mtp-humaneval26-crown-validate-20260528-144250/prompt.txt \
  --max-n 5 --trunk-spine --kv-mode q8 --temp 0.0 --max 256 --no-chatml
```

Notes:

- Prompt is HumanEval/26 `remove_duplicates`.
- Raw prompt md5: `90a10f2b41763aacc80a3e4db5e0bb62`.
- Normalized prompt md5 from the demo: `e04fc67ed919bec41bfd7426a0b75d58`.
- Published 5-run median: `259.53 tok/s`, tau `5.1250`, cycles `8`, committed `42`, eos `y`.
- This is the publish/crown MTP fixture. Do not substitute merge-sort/below-zero numbers for the crown claim.

### Dense 27B DFlash corrected-AWQ fixture

This is the corrected dense DFlash merge-sort fixture. Use this AWQ trunk only.

```bash
./target/release/examples/dflash_spec_demo \
  --target /home/kaden/.hipfire/models/qwen3.6-27b-awq.mq4 \
  --draft /home/kaden/.hipfire/models/qwen36-27b-dflash-mq4.hf4 \
  --prompt-file benchmarks/prompts/merge_sort_thinking_off.txt \
  --max 256 --temp 0.0 --no-chatml --kv-mode q8 --ctx 4096
```

Notes:

- Prompt md5: `253c7ac50857fe6d0e10fb0d2c5e35c0`.
- Do **not** use `/home/kaden/.hipfire/models/qwen3.6-27b-dev/qwen3.6-27b.mq4-awq-gptq-f2-lmhead-a100.hfq` for this fixture.
- This is the dense DFlash LocalMaxxing-style fixture, not the A3B DFlash research fixture.

### A3B MoE DFlash research fixture

Only these two prompts are permitted for ongoing A3B DFlash comparisons:

- `benchmarks/prompts/merge_sort_thinking_off.txt`
- `benchmarks/prompts/humaneval_3_below_zero.txt`

Use fixed small-B, not adaptive-B:

```bash
./target/release/examples/dflash_spec_demo \
  --target /home/kaden/.hipfire/models/qwen3.6-35b-a3b.mq4-awq-mi300x \
  --draft /home/kaden/.hipfire/models/qwen36-35b-a3b-dflash-mq4.hfq \
  --prompt-file benchmarks/prompts/merge_sort_thinking_off.txt \
  --max 256 --temp 0.0 --no-chatml --kv-mode q8 --ctx 4096 \
  --block-size 6 --no-adaptive-b
```

Swap only the prompt path to run the below-zero fixture:

```bash
--prompt-file benchmarks/prompts/humaneval_3_below_zero.txt
```

Notes:

- Best observed merge-sort shape on this branch was B=6 fixed, not adaptive-B.
- Use q8 KV for these comparisons.
- Treat A3B DFlash as research/diagnostic here; the branch's publishable A3B result is the MTP trunk-spine fixture above.
